WIP Add circe layer

aijony opened the pull request:
Fix issue #4231

After trying all 3 IRC clients (circe, rcirc, erc),
I enjoyed circe the most, so here is my first attempt at
contributing a layer. Hopefully it is up to par.

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,162 @@
+    (add-hook 'circe-channel-mode-hook #'enable-lui-autopaste)))
+
+(defun circe/post-init-helm ()
+  (use-package helm
```
the only place a `use-package` declaration should be located is the `init` function. Otherwise you are calling it twice.

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,162 @@
+      (setq lui-time-stamp-format "%H:%M")
+      (setq lui-fill-type nil)))
+
+  (use-package circe-color-nicks
```
All used packages need to be in the `circe-packages` variable. They also need to be wrapped in `circe/[post-]init-<package>` function. This applies to `circe-color-nicks`, `lui`, and `lui-autopaste.`

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,162 @@
+  (use-package lui
+    :init
+    (progn
+      (defun circe/set-local-vars ()
```
Use a `lambda` instead of a `defun` so you don't pollute your namespace with functions that you won't need again.

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,162 @@
+    :config
+    (setq circe-color-nicks-everywhere t))
+
+  (use-package lui
```
all packages that can be autoloaded should be autoloaded. Add something like `:commands lui-mode` to defer loading this package.

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,162 @@
+       (add-hook 'circe-mode-hook #'spacemacs-layouts/add-circe-buffer-to-persp)
+       (call-interactively 'circe)))
+
+  )
```
the extra whitespace and hanging parenthesis in `circe/post-init-persp-mode` is considered bad form.

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,162 @@
+
+      ;; Show channel name in prompt.
+
+      (defun circe/bufname-to-channame (s)
```
use lambda instead of defun for functions you only need once. If you want these function permanently add them to funcs.el.

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,162 @@
+
+  (use-package lui
+    :init
+    (progn
```
`:init` multiple forms so `progn` is not needed.

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,162 @@
+    :init
+    (add-hook 'circe-server-connected-hook #'enable-circe-notifications)
+    :config
+    (progn
```
`:config` accepts multiple forms so `progn` is not needed

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,162 @@
+  (use-package circe
+    :commands circe
+    :config
+    (progn
```
`:config` accepts multiple forms so `progn` is not needed

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,162 @@
+      (spacemacs/set-leader-keys "aic" 'circe)
+
+      (dolist (mode circe-modes)
+        (progn
```
the `dolist` body consists of multiple forms and does not need a `progn`.

aijony commented:
Thank you so much.  I'm sorry about my poor Elisp skills. I'll make sure to keep these tips in mind for future pull requests.

Hopefully, I got through all of those suggestions correctly and haven't added the need for new ones.
Your suggestions inspired some "suggestions" of my own too, so thank you for that.

- Removed all instances of `progn`
- Included `lui`, `lui-autopaste`, `circe-color-nicks` in `circe-packages` variable
- Implemented more instances of `lambda` and `let` instead of `defun`
- Helm no longer has `use-package`
- Added another configuration variable that I missed
- General refactors (hopefully with better form)
- Last minute fix on random typos

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,151 @@
+    circe-notifications
+    persp-mode
+    helm
+    helm-circe
```
all packages you use need to be in this list. You're still missing `circe-color-nicks`, `lui`, and `lui-autopaste`. if some of them rely on other packages to be loaded first you can use something like     `(<package> :toggle (configuration-layer/package-usedp '<required-package>))`

CeleritasCelery commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,151 @@
+
+(defun circe/init-helm-circe ()
+  (use-package helm-circe
+    :defer t
```
when you set `:defer t` it tells use-package that some other code is going to load the package. I don't think this is the case. Instead use the `:comands` keyword to tell use-package to load this package when certain commands are run, which will implicitly defer the package. So in this example replace `:defer t` with `:commands (helm-circe helm-circe-servers <etc>)`.

aijony commented:
Okay, I think I realized why I'm confused and why I'm confusing you.
Packages `circe-color-nicks`, `lui`, and `lui-autopaste` used to be optional for Circe.
Because of this, I was reading documentation that required them to be explicitly stated.
I don't know the technical term of what you would call them, but they are [included](https://github.com/jorgenschaefer/circe/blob/47bbcf4268602688f90c08ce64ec93226fc3f894/circe.el#L46-L47).
Please let me know if you think I am misunderstanding something.

Sorry about the issues, hopefully I've sorted everything out correctly.
Thank you for your kind patience.

robbyoconnor commented:
How in god's name did this not exist prior?

greboide commented:
amazing, when will we git this to develop?

Compro-Prasad commented:
This looks old. Is this getting into develop or is it having some issues?

Miciah commented on layers/+chat/circe/README.org:
```diff
@@ -0,0 +1,84 @@
+- [[#spacemacs-layout-support][Spacemacs Layout Support]]
+
+* Description
+Layer for [[http://www.emacswiki.org/emacs/Circe][Circe IRC chat]]. 
```
Needs a "the", and should say "client": "the Circe IRC client".

The wiki says that https://github.com/jorgenschaefer/circe is the homepage for Circe.  Is there a reason to link to the wiki instead of the GitHub page?

Miciah commented on layers/+chat/circe/README.org:
```diff
@@ -0,0 +1,84 @@
+* Spacemacs Layout Support
+CIRCE buffers are added automatically to the default layout for CIRCE controlled by
+the variable ~circe-spacemacs-layout-name~ (defaults to "@circe") and the default
+binding ("i") is controlled by ~circe-spacemacs-layout-binding~. 
```
Use \`, not ~, for variable names.

Miciah replied:
Sorry, \` would also be wrong; use = for variables.

robbyoconnor replied:
You should look at how other layers do it and work from there.

Miciah commented on layers/+chat/circe/config.el:
```diff
@@ -0,0 +1,24 @@
+  "List of modes that are used by circe")
+
+(defvar circe-helm-for-username-completion 'nil
+  "Whether or not to use helm for username completion")
```
Why is this `'nil` instead of `nil`? (Related, why does the documentation say to use `'t` instead of `t`?)

Miciah commented on layers/+chat/circe/config.el:
```diff
@@ -0,0 +1,24 @@
+;; Variables
+
+(defvar circe-spacemacs-layout-name "@circe"
+  "Name used in the setup for `spacemacs-layouts' micro-state")
```
This and the other docstrings need full stops.

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+    (setq circe-format-say "{nick}> {body}")
+    (setq circe-format-self-say (concat (propertize ">>>" 'face 'circe-self-say-face) " {body}"))
+    (setq circe-prompt-string (concat (propertize ">>>" 'face 'circe-prompt-face) " "))
+    (setq circe-highlight-nick-type 'all)
```
Can the `setq` forms be moved to `:init`? That way, users can set them in `dotspacemacs/user-config`.  By setting them in `:config`, you overwrite any user-specified settings.

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+
+    ;; Add circe-color-nicks
+    (enable-circe-color-nicks)
+    (setq circe-color-nicks-everywhere 't)
```
Can `'t` be just `t`?

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+    ;; Show channel name in prompt.
+    (add-hook 'circe-chat-mode-hook
+              (lambda ()
+                (-let* (
```
Sylvain has said elsewhere that we want to avoid using dash (see https://github.com/syl20bnr/spacemacs/pull/10792#discussion_r192637508).

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+    (setq lui-time-stamp-position 'right-margin)
+    (setq lui-time-stamp-format "%H:%M")
+    (setq lui-fill-type nil)
+    (add-hook 'lui-mode-hook (lambda ()
```
Better to add a named function to the layer's `funcs.el` and use that instead of a lambda.

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+  (use-package helm-circe
+    :commands (helm-circe
+               helm-circe-servers
+               helm-circe-queries
```
You shouldn't need to specify `:commands`; all of these commands are in the helm-circe package's autoloads.

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+        "h q" 'helm-circe-queries
+        "h b" 'helm-circe-by-server
+        "h n" 'helm-circe-new-activity
+        ))))
```
These last two lines should be joined so that the parentheses hug the last line.

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+
+(defun circe/init-circe-notifications ()
+  (use-package circe-notifications
+    :commands enable-circe-notifications
```
The `:commands` should be unnecessary because the circe-notifications package specifies the autoload for `enable-circe-notifications`.

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+
+(defun circe/post-init-persp-mode ()
+  ;; do not save circe buffers
+  (with-eval-after-load 'persp-mode
```
Other layers use `(spacemacs|use-package-add-hook persp-mode :post-config ...)`.

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+  ;; do not save circe buffers
+  (with-eval-after-load 'persp-mode
+    (push (lambda (b) (with-current-buffer b (eq major-mode 'circe-mode)))
+          persp-filter-save-buffers-functions))
```
Better to add a named function to `funcs.el` and use that instead of a lambda.

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+  (with-eval-after-load 'helm
+    (if (not circe-helm-for-username-completion)
+        (dolist (mode circe-modes)
+          (add-to-list 'helm-mode-no-completion-in-region-in-modes mode)))))
```
I recommend using `when` instead of `if` when there is no `else` form.

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+        "c h" 'circe-command-HELP
+        "c t" 'circe-command-CHTOPIC
+        "c s" 'circe-command-SAY
+        ))))
```
These last two lines should be joined so that the parentheses hug the last line.

Miciah commented on layers/+chat/circe/packages.el:
```diff
@@ -0,0 +1,142 @@
+
+    ;; Show channel name in prompt.
+    (add-hook 'circe-chat-mode-hook
+              (lambda ()
```
Better to add a named function to the layer's `funcs.el` and use that named function here.

aijony commented:
Thanks for all the the tips @Miciah, I'm still very much a beginner to elisp. I was just playing around with circe at the time, wrote a layer for myself, and thought it decent enough to share. I have now moved back to erc,
so I haven't touched this layer at all. When I get the chance, I will apply those wonderful comments.

duianto commented:
I added the prefix `WIP` to make it clear from the pull request list that this isn't ready.
Remove the prefix when it is ready.
